go/lru/disk_lru.go

```diff
@@ -0,0 +1,320 @@
+package lru
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/libkb"
+	context "golang.org/x/net/context"
+)
+
+type DiskLRUEntry struct {
+	Key          string
+	Value        interface{}
+	Ctime        time.Time
+	LastAccessed time.Time
+}
+
+type diskLRUIndex struct {
+	sync.Mutex
+	Version     int
+	EntryKeys   []string
+	entryKeyMap map[string]bool
+	dirty       bool
+}
+
+func newDiskLRUIndex(version int) *diskLRUIndex {
+	return &diskLRUIndex{
+		Version: version,
+	}
+}
+
+func (d *diskLRUIndex) mapify() {
+	if d.entryKeyMap == nil {
+		d.entryKeyMap = make(map[string]bool)
+		for _, k := range d.EntryKeys {
+			d.entryKeyMap[k] = true
+		}
+	}
+}
+
+func (d *diskLRUIndex) exists(key string) bool {
+	d.mapify()
+	return d.entryKeyMap[key]
+}
+
+func (d *diskLRUIndex) Exists(key string) bool {
+	d.Lock()
+	defer d.Unlock()
+	return d.exists(key)
+}
+
+func (d *diskLRUIndex) remove(key string) {
+	if d.entryKeyMap != nil {
+		delete(d.entryKeyMap, key)
+	}
+	var ne []string
+	for _, k := range d.EntryKeys {
```

Comment:
are you worried about this linear scan?

Reply:
Seems like you hit it whenever you access anything. Not only that, but you need to recook the whole list, is that going to make a swisscheese of memory?

Reply:
It's not ideal, but I was looking at a use case with maybe 1000 items in here, so I thought I could get away with something simple and double back if it is a problem.